packages/yarnpkg-pnpify/sources/Hoister.ts

```diff
@@ -0,0 +1,289 @@
+  pnpifyFs?: boolean;
+}
+
+/** node_modules path segment */
+const NODE_MODULES = toFilename(`node_modules`);
+const NODE_MODULES_SUFFIX = `/node_modules`;
+
+/** Package locator key for usage inside maps */
+type LocatorKey = string;
+
+/**
+ * Retrieves full package list from PnP API and builds hoisted `node_modules` directories
+ * representation in-memory.
+ */
+export class Hoister {
+  /** Raw hoister which does all the hoisting heavy lifting */
+  private readonly rawHoister = new RawHoister();
+
+  /** Hoister options */
+  private readonly options: HoisterOptions;
+
+  public constructor(options: HoisterOptions = {optimizeSizeOnDisk: false}) {
+    this.options = options;
+  }
+
+  /**
+   * Traverses PnP tree and produces input for the `RawHoister`
+   *
+   * @param pnp PnP API
+   *
+   * @returns package tree, packages info and locators
+   */
+  private buildPackageTree(pnp: PnpWalkApi): { packageTree: ReadonlyPackageTree, packages: PackageInfo[], locators: PackageLocator[] } {
+    const packageTree: PackageTree = [];
+    const packages: PackageInfo[] = [];
+    const locators: PackageLocator[] = [];
+    const locatorToPackageMap = new Map<LocatorKey, PackageId>();
+    const packageInfos: PackageInformation<NativePath>[] = [];
+
+    const pnpRoots = pnp.getDependencyTreeRoots ? pnp.getDependencyTreeRoots() : [pnp.topLevel];
+
+    let lastPkgId = 0;
+
+    const getLocatorKey = (locator: PackageLocator): LocatorKey => `${locator.name}:${locator.reference}`;
+
+    const assignPackageId = (locator: PackageLocator, pkg: PackageInformation<NativePath>) => {
+      const locatorKey = getLocatorKey(locator);
+      let pkgId = locatorToPackageMap.get(locatorKey);
+      if (!pkgId) {
+        pkgId = lastPkgId;
+        lastPkgId++;
+        const packagePath = npath.toPortablePath(pkg.packageLocation);
+        let weight = 1;
+        if (this.options.optimizeSizeOnDisk && packagePath.indexOf(`.zip${NODE_MODULES_SUFFIX}`) > 0)
+          weight = fs.statSync(packagePath.split(NODE_MODULES_SUFFIX)[0]).size;
+
+        locatorToPackageMap.set(locatorKey, pkgId);
+        locators.push(locator);
+        packages.push({name: locator.name!, weight});
+        packageInfos.push(pkg);
+      }
+      return pkgId;
+    };
+
+    const getPackageId = (locator: PackageLocator) => locatorToPackageMap.get(getLocatorKey(locator));
+
+    const addedIds = new Set<PackageId>();
+    const addPackageToTree = (pkg: PackageInformation<NativePath>, pkgId: PackageId, parentDepIds: Set<PackageId>) => {
+      addedIds.add(pkgId);
+      const deps = new Set<PackageId>();
+      const peerDeps = new Set<PackageId>();
+      packageTree[pkgId] = {deps, peerDeps};
+
+      for (const [name, reference] of pkg.packageDependencies) {
+        if (reference) {
+          const locator = typeof reference === 'string' ? {name, reference} : {name: reference[0], reference: reference[1]};
+          const depPkgId = getPackageId(locator);
```

Comment:
In which case can a package ID be missing? 🤔

Reply:
Removed check, not sure why I added it